### DISCUSSION TITLE
Ask us anything

### DISCUSSION THREAD

**Discussion by peterwaller-arm** [elfshaker/elfshaker#58] [Q&A] [answered]
We're here, online for the next few hours. The project is new, there will be bugs, things will not be clear -- ask questions, file issues, and help us make this even more great. Thanks!

**Comment by rovashan** [comment]
Apart from version control, what are some of the use cases for Elfshaker?

**Reply by peterwaller-arm** [reply]
Our immediate use case is the [manyclangs](https://github.com/elfshaker/manyclangs/) project, where elfshaker enables us to share 100MiB pack files which contain many (thousands) of builds of clang, which would ordinarily cost hundreds of megabytes each.

Another use case I have in mind is to capture the state of your build at every moment when you're building complex software. Then you can potentially go back to a previous state quickly. However, ccache might also be quite suitable for this, and it's early days for the project.

There might be other use cases out there that we don't yet know about, let us know if you find something neat!

**Comment by nh2** [comment]
Extracting question from https://news.ycombinator.com/item?id=29277494:

How does elfshaker compare to other version-controlling, packfile-deduplicating programs / algorithms like the ones in:

* [bup](https://github.com/bup/bup/) ([design](https://github.com/bup/bup/blob/9118aca5f232cc9a4da573a367da31ea8bee2fd5/DESIGN#L64))
* [kopia](https://github.com/kopia/kopia)
* [restic](https://restic.net)
* [borg](https://www.borgbackup.org/)
* [bupstash](https://bupstash.io/) (also written in Rust)

Did you know of their existence when making elfshaker? If yes, it would be cool to learn about design tradeoffs.

**Reply by peterwaller-arm** [reply]
Before starting out some time ago, I did some experiments with bup. I had a good experience with bup and high expectations for it. However, I found that quite a lot of performance was left on the table, so I was motivated to start elfshaker. Unfortunately that time has past so I don't have scientific numbers for you measured with other software at this time.

As an idea of how elfshaker performs, we see ~300ms time to create a snapshot for clang, and ~seconds-to-minute to create a binary pack containing thousands of revisions. Extraction takes less than a second. One difference of elfshaker compared with some other software I tested is that we do the compression and decompression in parallel, which can make a very big difference on today's many-core machines.

**Comment by svilendobrev** [comment]
will this work well for compressed / uncompressed audio files? mp3, flac, wav ..

**Reply by veselink1** [reply]
I haven't tested, but my guess is no. elfshaker excels when files across snapshots don't change often and when the diff between versions of these files is small when they do change.

mp3 files in particular (might be true for flac as well) are already compressed, so revisions of these files will compress poorly. 

We just added a section on [applicability](https://github.com/elfshaker/elfshaker#applicability) in the README.

**Comment by samsieber** [comment]
From what I understand this requires tweaks to how object files are laid out. Does this have any performance impact? And do you see something like this being doable with binaries from other languages, e.g. Rust?

**Reply by veselink1** [reply]
elfshaker, despite its name, does not alter object files. At its core, it is a relatively simple version control system with a content-addressable storage model and support for compressed pack files. It is similar to git in that way, only our tests show that it is much more performant and has higher compression ratios. We attribute this to our use of Zstandard, and our [pack format](https://github.com/elfshaker/elfshaker/blob/main/docs/contributors/format.md#pack-format).

elfshaker can be used to store and pack arbitrary data - the compression ratio will depend on the type of data being stored. Most of the size reduction in our results comes from deduplication. The other size-reduction step we use is Zstandard, which has about 10% compression ratio on the deduplicated set of files. The format we use has been tuned for use in [manyclangs](https://github.com/elfshaker/manyclangs).

Generally, if you know you have a use-case that involves many duplicate files across snapshots, and you want fast access to these files, elfshaker should be a good fit. I would caution against using elfshaker for sensitive data you cannot recover.

**Comment by abgoyal** [comment]
If I build a pack file with, say 200 builds of clang, and share the pack file with a friend - and then later, if I update my pack file to include a 201st build, is there a way for my friend to get this updated pack file without sending the whole thing over? Meaning can the pack files be delta-synced efficiently with others?

**Reply by peterwaller-arm** [reply]
This is not currently implemented, but should be possible in principle. For simplicity of implementation my first plan was just to not bother with delta syncing and instead just send the whole fresh packfile, since a 2kbuild packfile for a month only amounts to 100MiB. Therefore, the effort of implementing delta syncing, which might be nice to have, may not provide a huge benefit compared to what is already possible.

**Reply by veselink1** [reply]
There is no built-in support for this use-case at the moment.

Snapshots (builds in your case) created by `elfshaker store` are referred to as 'loose'. The files belonging to those snapshots are not packed (not in a pack file, not compressed). A number of snapshots can be packed by `elfshaker pack` into a pack file, which compresses them. 

Currently, elfshaker can only create new packs. Updating pack files is tracked in https://github.com/elfshaker/elfshaker/issues/35. Once #35 is implemented, you will be able to add snapshots to an existing pack. 

Delta-syncing, which I interpret as meaning sending only the files not present on the receiving end, is not implemented. What you could do instead, is to create another pack, containing only your 201st build (or N builds) and send it over to your friend. Your friend will then be able to run the repacking command from #35, to update their own pack.

**Comment by shi-yan** [comment]
so how do I do diff and merge (resolve conflict) ? possible to do branching?

**Reply by veselink1** [reply]
We do not support these operations. Our use case for elfshaker is binary files, for which these manual operations make less sense I believe. We do not use compute file deltas in our storage model in general. Each pack file in elfshaker maintains its own independent linear history of snapshots, so no branching either.

[File Format](https://github.com/elfshaker/elfshaker/blob/main/docs/contributors/format.md)
[Usage Guide](https://github.com/elfshaker/elfshaker/blob/main/docs/users/usage.md)

**Comment by hongtaihu** [comment]
Hi! The current aarch64-ubuntu2004.esi manifest seems to provide packs only through 202605.
  Is the 202606 pack delayed, or has monthly publication paused?
  Thanks!

**Reply by peterwaller-arm** [reply]
Hi @hongtaihu - thanks for the ping, it's delayed. I plan to publish it imminently (by tomorrow), and then get back to regular updates.

**Reply by hongtaihu** [reply]
Thanks for the update!

**Reply by peterwaller-arm** [reply]
It's up now. Feel free to ping again in the future if things aren't available.

**Comment by hongtaihu** [comment]
My codex says:
```
aarch64-ubuntu2004-manyclangs-202403.pack cannot be extracted. `list` works (1941 snapshots). `extract --reset` of
  any 202403 snapshot fails with `Corrupted block detected`.

  Release asset size matches GitHub: 139333315 bytes, sha256
  `73061313889a84e30b11c3f766946704ca9715df67baf8d5aecdedbf50ab5e6a`.

  Same elfshaker v0.9.0: 202402 and 202404 extract fine.

  libzstd 1.5.5 (`windowLogMax=31`): first frame at offset 241 (`28b52ffd`) dies with `Data corruption detected` after
  ~982KB in / 23MB out. 202402 (17 frames) and 202404 (21 frames) fully decompress.
```
I am not sure whether it happens.

**Reply by peterwaller-arm** [reply]
The sha256 that I measure on github for https://github.com/elfshaker/manyclangs/releases/download/v0.9.0/aarch64-ubuntu2004-manyclangs-202403.pack is `d823f94805be6b84867e1dbe9b84a9bc9e365bef63dc92ae616e82215878e8a7`. Extracting from this one works.

Can you re-download and remeasure?

**Reply by hongtaihu** [reply, marked as answer]
> The sha256 that I measure on github for https://github.com/elfshaker/manyclangs/releases/download/v0.9.0/aarch64-ubuntu2004-manyclangs-202403.pack is `d823f94805be6b84867e1dbe9b84a9bc9e365bef63dc92ae616e82215878e8a7`. Extracting from this one works.
> 
> Can you re-download and remeasure?

OK, it was the AI that made the mistake. Nothing needs to be fixed.

**Reply by peterwaller-arm** [reply]
Thanks for taking the time to report back!

**Reply by peterwaller-arm** [reply]
Additionally, someone else identified some incorrect checksums specified in the esi which have been corrected: https://github.com/elfshaker/manyclangs/issues/17 - I don't think that is the problem you faced but worth noting here.

**Comment by hongtaihu** [comment]
On Commit: 85043c1c146fd5658ad4c5b5138e58994333e645 in aarch64-ubuntu2004-manyclangs-202508

  Input:
```cxx
    void test() {}
```
``` 
   clang++ -fsyntax-only -std=c++23 test.cpp
```
  I found Behavior:
    manyclangs snapshot of this commit crashes while a native build of the same commit does not. Does this only happen on my PC and is this expected?

**Reply by peterwaller-arm** [reply]
I have reproduced this locally and I believe it is manyclangs build issue we want to fix which is not yet fully understood. Thanks for the report. Please file an issue in the repository. Meanwhile it seems just that commit is crashy, you can `git bisect skip` it (unless it is the implicated commit in which case your native build will help you). Reports like this are useful though because if there are problems with the build process then we want to get concrete reports of them and fix them over time. This is particularly timely as there is renewed interest in this at the moment.
